File: labs/11-lsp-minimum/java/Main.java
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Main {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final Map<String, String> DOCUMENTS = new ConcurrentHashMap<>();

    public static void main(String[] args) throws Exception {
        if (args.length > 0 && "--smoke".equals(args[0])) {
            smokeTest();
        } else {
            serve(System.in, System.out);
        }
    }

    private static void serve(InputStream in, OutputStream out) throws Exception {
        System.err.println("LSP server ready on stdio");
        while (true) {
            JsonNode request;
            try {
                request = readMessage(in);
            } catch (Exception exc) {
                System.err.println("Bad JSON: " + exc.getMessage());
                writeMessage(out, makeError(null, -32700, "Parse error"));
                continue;
            }
            if (request == null) {
                System.err.println("EOF reached; shutting down");
                break;
            }
            String method = request.get("method").asText();
            JsonNode response = handleRequest(request);
            if (response != null) {
                writeMessage(out, response);
            }
            if ("exit".equals(method)) {
                break;
            }
        }
    }

    private static JsonNode readMessage(InputStream in) throws Exception {
        int contentLength = 0;
        while (true) {
            String line = readLine(in);
            if (line == null) {
                return null;
            }
            line = line.trim();
            if (line.isEmpty()) {
                break;
            }
            if (line.contains(":")) {
                String[] parts = line.split(":", 2);
                if ("content-length".equalsIgnoreCase(parts[0].trim())) {
                    contentLength = Integer.parseInt(parts[1].trim());
                }
            }
        }
        if (contentLength == 0) {
            return null;
        }
        byte[] body = new byte[contentLength];
        int read = 0;
        while (read < contentLength) {
            int n = in.read(body, read, contentLength - read);
            if (n < 0) {
                return null;
            }
            read += n;
        }
        return MAPPER.readTree(new String(body, StandardCharsets.UTF_8));
    }

    private static String readLine(InputStream in) throws IOException {
        ByteArrayOutputStream buf = new ByteArrayOutputStream();
        int b;
        while ((b = in.read()) != -1) {
            if (b == '\n') {
                break;
            }
            if (b != '\r') {
                buf.write(b);
            }
        }
        if (b == -1 && buf.size() == 0) {
            return null;
        }
        return buf.toString(StandardCharsets.UTF_8);
    }

    private static void writeMessage(OutputStream out, JsonNode message) throws Exception {
        byte[] body = MAPPER.writeValueAsBytes(message);
        out.write(("Content-Length: " + body.length + "\r\n\r\n").getBytes(StandardCharsets.UTF_8));
        out.write(body);
        out.flush();
    }

    private static JsonNode handleRequest(JsonNode request) {
        String method = request.get("method").asText();
        JsonNode params = request.get("params");
        if (params == null) {
            params = MAPPER.createObjectNode();
        }
        JsonNode reqId = request.get("id");

        System.err.println("Received " + method);

        return switch (method) {
            case "initialize" -> makeResponse(reqId, MAPPER.createObjectNode()
                    .set("capabilities", MAPPER.createObjectNode()
                            .set("textDocumentSync", MAPPER.createObjectNode().put("openClose", true).put("change", 0))
                            .put("definitionProvider", true))
                    .set("serverInfo", MAPPER.createObjectNode().put("name", "agent-labs-lsp").put("version", "0.1.0")));
            case "initialized" -> null;
            case "textDocument/didOpen" -> {
                JsonNode textDocument = params.get("textDocument");
                updateDocument(textDocument.get("uri").asText(), textDocument.get("text").asText());
                yield null;
            }
            case "textDocument/definition" -> {
                JsonNode textDocument = params.get("textDocument");
                JsonNode position = params.get("position");
                JsonNode location = findDefinition(textDocument.get("uri").asText(), position);
                yield makeResponse(reqId, location);
            }
            case "shutdown" -> makeResponse(reqId, null);
            case "exit" -> null;
            default -> makeError(reqId, -32601, "Method not found: " + method);
        };
    }

    private static JsonNode makeResponse(JsonNode id, JsonNode result) {
        ObjectNode response = MAPPER.createObjectNode();
        response.put("jsonrpc", "2.0");
        response.set("id", id);
        response.set("result", result);
        return response;
    }

    private static JsonNode makeError(JsonNode id, int code, String message) {
        ObjectNode response = MAPPER.createObjectNode();
        response.put("jsonrpc", "2.0");
        response.set("id", id);
        ObjectNode error = response.putObject("error");
        error.put("code", code);
        error.put("message", message);
        return response;
    }

    private static void updateDocument(String uri, String text) {
        DOCUMENTS.put(uri, text);
    }

    private static JsonNode findDefinition(String uri, JsonNode position) {
        String text = DOCUMENTS.getOrDefault(uri, "");
        String[] lines = text.split("\n", -1);
        int lineIdx = position.has("line") ? position.get("line").asInt() : 0;
        int charIdx = position.has("character") ? position.get("character").asInt() : 0;
        if (!(0 <= lineIdx && lineIdx < lines.length)) {
            return null;
        }
        String line = lines[lineIdx];
        String after = charIdx < line.length() ? line.substring(charIdx) : "";
        Matcher matcher = Pattern.compile("[A-Za-z0-9_]+").matcher(after);
        if (!matcher.find()) {
            return null;
        }
        String word = matcher.group(0);
        Pattern defPattern = Pattern.compile("^def\\s+" + Pattern.quote(word) + "\\s*\\(", Pattern.MULTILINE);
        for (Map.Entry<String, String> entry : DOCUMENTS.entrySet()) {
            Matcher defMatcher = defPattern.matcher(entry.getValue());
            if (defMatcher.find()) {
                int startLine = entry.getValue().substring(0, defMatcher.start()).split("\n", -1).length - 1;
                ObjectNode location = MAPPER.createObjectNode();
                location.put("uri", entry.getKey());
                ObjectNode range = location.putObject("range");
                ObjectNode start = range.putObject("start");
                start.put("line", startLine);
                start.put("character", 0);
                ObjectNode end = range.putObject("end");
                end.put("line", startLine);
                end.put("character", ("def " + word + "(").length());
                return location;
            }
        }
        return null;
    }

    private static void smokeTest() throws Exception {
        String sampleCode = String.join("\n", List.of(
                "def greet(name):",
                "    return f\"Hello, {name}!\"",
                "",
                "print(greet(\"world\"))"
        ));
        String uri = "file:///tmp/sample.py";

        List<JsonNode> requests = List.of(
                makeRequest(1, "initialize", MAPPER.createObjectNode()
                        .put("processId", (String) null).set("rootUri", null).set("capabilities", MAPPER.createObjectNode())),
                makeRequest(null, "initialized", MAPPER.createObjectNode()),
                makeRequest(null, "textDocument/didOpen", MAPPER.createObjectNode()
                        .set("textDocument", MAPPER.createObjectNode()
                                .put("uri", uri)
                                .put("languageId", "python")
                                .put("text", sampleCode))),
                makeRequest(2, "textDocument/definition", MAPPER.createObjectNode()
                        .set("textDocument", MAPPER.createObjectNode().put("uri", uri))
                        .set("position", MAPPER.createObjectNode().put("line", 3).put("character", 6))),
                makeRequest(3, "shutdown", MAPPER.createObjectNode()),
                makeRequest(null, "exit", MAPPER.createObjectNode())
        );

        ByteArrayOutputStream stdin = new ByteArrayOutputStream();
        for (JsonNode req : requests) {
            byte[] body = MAPPER.writeValueAsBytes(req);
            stdin.write(("Content-Length: " + body.length + "\r\n\r\n").getBytes(StandardCharsets.UTF_8));
            stdin.write(body);
        }
        ByteArrayInputStream in = new ByteArrayInputStream(stdin.toByteArray());
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        serve(in, out);
        System.out.println("Smoke test output:");
        System.out.println(out.toString(StandardCharsets.UTF_8));
    }

    private static JsonNode makeRequest(Integer id, String method, JsonNode params) {
        ObjectNode req = MAPPER.createObjectNode();
        req.put("jsonrpc", "2.0");
        if (id != null) {
            req.put("id", id);
        } else {
            req.set("id", null);
        }
        req.put("method", method);
        req.set("params", params);
        return req;
    }
}

File: labs/11-lsp-minimum/node/main.ts
import { PassThrough, Readable } from "stream";

const DOCUMENTS: Record<string, string> = {};

function updateDocument(uri: string, text: string): void {
  DOCUMENTS[uri] = text;
}

function findDefinition(uri: string, position: any): any | null {
  const text = DOCUMENTS[uri] || "";
  const lines = text.split("\n");
  const lineIdx = position.line || 0;
  const charIdx = position.character || 0;
  if (lineIdx < 0 || lineIdx >= lines.length) return null;
  const line = lines[lineIdx];
  const match = (line.slice(charIdx) || "").match(/[A-Za-z0-9_]+/);
  if (!match) return null;
  const word = match[0];

  const pattern = new RegExp(`^def\\s+${word.replace(/[.*+?^${}()|[\]\\]/g, "\\$\u0026")}\\s*\\(`, "m");
  for (const docUri of Object.keys(DOCUMENTS)) {
    const docText = DOCUMENTS[docUri];
    const m = docText.match(pattern);
    if (m) {
      const startLine = docText.slice(0, m.index).split("\n").length - 1;
      return {
        uri: docUri,
        range: {
          start: { line: startLine, character: 0 },
          end: { line: startLine, character: `def ${word}(`.length },
        },
      };
    }
  }
  return null;
}

// ---------- JSON-RPC framing ----------

class StreamReader {
  private buffer = Buffer.alloc(0);
  private waiter: (() => void) | null = null;
  private ended = false;

  constructor(stream: NodeJS.ReadableStream) {
    stream.on("data", (chunk: Buffer) => {
      this.buffer = Buffer.concat([this.buffer, chunk]);
      if (this.waiter) {
        this.waiter();
        this.waiter = null;
      }
    });
    stream.on("end", () => {
      this.ended = true;
      if (this.waiter) {
        this.waiter();
        this.waiter = null;
      }
    });
  }

  private async wait(): Promise<void> {
    if (this.buffer.length > 0 || this.ended) return;
    await new Promise<void>((resolve) => {
      this.waiter = resolve;
    });
  }

  async readUntil(sep: string): Promise<Buffer | null> {
    const sepBuf = Buffer.from(sep);
    while (true) {
      const idx = this.buffer.indexOf(sepBuf);
      if (idx !== -1) {
        const result = this.buffer.slice(0, idx);
        this.buffer = this.buffer.slice(idx + sepBuf.length);
        return result;
      }
      if (this.ended) return this.buffer.length > 0 ? this.buffer : null;
      await this.wait();
    }
  }

  async readExactly(n: number): Promise<Buffer | null> {
    while (this.buffer.length < n) {
      if (this.ended) return null;
      await this.wait();
    }
    const result = this.buffer.slice(0, n);
    this.buffer = this.buffer.slice(n);
    return result;
  }
}

async function* readMessages(reader: StreamReader): AsyncGenerator<any> {
  while (true) {
    const headerBytes = await reader.readUntil("\r\n\r\n");
    if (!headerBytes) break;
    const headers: Record<string, string> = {};
    for (const line of headerBytes.toString("utf8").split("\r\n")) {
      const colon = line.indexOf(":");
      if (colon !== -1) {
        headers[line.slice(0, colon).trim().toLowerCase()] = line
          .slice(colon + 1)
          .trim();
      }
    }
    const length = parseInt(headers["content-length"] || "0", 10);
    if (length === 0) break;
    const body = await reader.readExactly(length);
    if (!body) break;
    yield JSON.parse(body.toString("utf8"));
  }
}

function writeMessage(stream: NodeJS.WritableStream, message: any): void {
  const body = Buffer.from(JSON.stringify(message), "utf8");
  stream.write(`Content-Length: ${body.length}\r\n\r\n`);
  stream.write(body);
  (stream as any).flush?.();
}

// ---------- request handlers ----------

function makeResponse(requestId: any, result: any): any {
  return { jsonrpc: "2.0", id: requestId, result };
}

function makeError(requestId: any, code: number, message: string): any {
  return { jsonrpc: "2.0", id: requestId, error: { code, message } };
}

function handleRequest(request: any): any | null {
  const method = request.method;
  const params = request.params || {};
  const reqId = request.id;

  console.error("Received", method);

  if (method === "initialize") {
    return makeResponse(reqId, {
      capabilities: {
        textDocumentSync: { openClose: true, change: 0 },
        definitionProvider: true,
      },
      serverInfo: { name: "agent-labs-lsp", version: "0.1.0" },
    });
  }

  if (method === "initialized") return null;

  if (method === "textDocument/didOpen") {
    const doc = params.textDocument || {};
    updateDocument(doc.uri || "", doc.text || "");
    return null;
  }

  if (method === "textDocument/definition") {
    const td = params.textDocument || {};
    const pos = params.position || {};
    const location = findDefinition(td.uri || "", pos);
    return makeResponse(reqId, location);
  }

  if (method === "shutdown") return makeResponse(reqId, null);

  if (method === "exit") return null;

  return makeError(reqId, -32601, `Method not found: ${method}`);
}

async function serve(
  stdin: NodeJS.ReadableStream = process.stdin,
  stdout: NodeJS.WritableStream = process.stdout
): Promise<void> {
  const reader = new StreamReader(stdin);
  console.error("LSP server ready on stdio");
  for await (const request of readMessages(reader)) {
    const method = request.method;
    const response = handleRequest(request);
    if (response !== null) {
      writeMessage(stdout, response);
    }
    if (method === "exit") break;
  }
}

// ---------- smoke test ----------

async function smokeTest(): Promise<void> {
  const sampleCode = [
    'def greet(name):',
    '    return f"Hello, {name}!"',
    '',
    'print(greet("world"))',
  ].join("\n");
  const uri = "file:///tmp/sample.py";

  const requests = [
    {
      jsonrpc: "2.0",
      id: 1,
      method: "initialize",
      params: { processId: null, rootUri: null, capabilities: {} },
    },
    { jsonrpc: "2.0", method: "initialized", params: {} },
    {
      jsonrpc: "2.0",
      method: "textDocument/didOpen",
      params: {
        textDocument: { uri, languageId: "python", text: sampleCode },
      },
    },
    {
      jsonrpc: "2.0",
      id: 2,
      method: "textDocument/definition",
      params: {
        textDocument: { uri },
        position: { line: 3, character: 6 },
      },
    },
    { jsonrpc: "2.0", id: 3, method: "shutdown" },
    { jsonrpc: "2.0", method: "exit" },
  ];

  const chunks: Buffer[] = [];
  for (const req of requests) {
    const body = Buffer.from(JSON.stringify(req), "utf8");
    chunks.push(Buffer.from(`Content-Length: ${body.length}\r\n\r\n`));
    chunks.push(body);
  }
  const stdin = Readable.from(chunks);
  const stdout = new PassThrough();
  const output: Buffer[] = [];
  stdout.on("data", (c) => output.push(c));

  await serve(stdin, stdout);
  console.log("Smoke test output:");
  console.log(Buffer.concat(output).toString("utf8"));
}

async function main() {
  if (process.argv.includes("--smoke")) {
    await smokeTest();
  } else {
    await serve();
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
